File: chp_1/chapter_one.py
import math


def compare_insertion_with_merge_sort():
    n: int = 2
    insert_less_than_merge = True
    while insert_less_than_merge:
        insert_sort_steps = calc_insert_sort(n)
        merge_sort_steps = calc_merge_sort(n)

        insert_less_than_merge = insert_sort_steps < merge_sort_steps
        n += 1

    print(f"Insert Sort takes {insert_sort_steps} many steps to execute on {n} many inputs.")
    print(f"Merge Sort takes {merge_sort_steps} many steps to execute {n} many inputs.")


def calc_insert_sort(n: int) -> float:
    return 8 * n ** 2


def calc_merge_sort(n: int) -> float:
    return 64 * n * math.log2(n)


def smallest_value_n():
    n_sq_less_than_two_n: bool = True
    n: int = 1
    # n_sq_runtime = 0.0
    # two_power_n_runtime = 0.0
    while n_sq_less_than_two_n:
        # n_sq_runtime = 100 * n ** 2
        # two_power_n_runtime = 2 ** n
        if 100 * n ** 2 < 2 ** n: break
        n += 1
    # print(n_sq_runtime)
    # print(two_power_n_runtime)
    print(f"The smallest n where nSquared runtime is faster than 2PowerN is: {n}")



if __name__ == '__main__':
    compare_insertion_with_merge_sort()
    smallest_value_n()

File: chp_1/ChapterOne.java
package chp_1;

import java.lang.Math;
public class ChapterOne{
    /**
     * log2 N = loge N / loge 2
     *      Java doesn't have a native log2. To calculate log2 of N one must do this indirectly.
     *      Natural Log N divided by natural log 2.
     * */
    public static void compareInsertionSortWithMerge(){
        /* Suppose that for inputs of size n on a particular computer, insertion sort runs in 8n^2 steps and merge sort
            runs in 64 nlg n steps. For which value n does insertion sort beat merge sort.
        * */
        double n = 2;
        boolean insertLessThanMerge = true;
        double insertSortN = 0;
        double mergeSortN = 0;
        while(insertLessThanMerge){

        insertSortN = calculateInsertSort(n);
        mergeSortN = calculateMergeSort(n);
            insertLessThanMerge= insertSortN < mergeSortN;
            n++;
        }

        System.out.println("Insert Sort takes "+ insertSortN +" many steps to execute on " + n + " many inputs.");
        System.out.println("Merge Sort takes "+ mergeSortN +" many steps to execute " + n + " many inputs.");
    // 45
    }

    private static double calculateInsertSort(double n){
        return 8 * Math.pow(n,2);
    }
    private static double calculateMergeSort(double n){
        return 64 * n * (Math.log(n)/Math.log(2));
    }

    public static void smallestValueOfN(){
        // Smallest value of n such at an algorithm whose running time is 100n2 runs faster than an algorithm whose
        //  running time is 2n

        boolean nSquaredLessThanTwoN = false;
        int n = 1;
        double nSquareRunTime = 0;
        double twoPowerNRunTime = 0;
        while(!nSquaredLessThanTwoN){

            nSquareRunTime = 100 * Math.pow(n, 2);
            twoPowerNRunTime = Math.pow(2, n);
            if(nSquareRunTime < twoPowerNRunTime){break;}
            ++n;
        }
        System.out.println(nSquareRunTime);
        System.out.println(twoPowerNRunTime);
        System.out.println("The smallest n where nSquared runtime is faster than 2PowerN is: " + n);

    }
}